File: Programmers/Backtracking/피로도.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;
void backtracking(int k, vector<vector<int>> dungeons, vector<int> visited, int cnt) {
    answer = max(answer, cnt);

    for (int i = 0; i < dungeons.size(); i++) {
        if (k >= dungeons[i][0] && visited[i] == 0) { // 최소 피로도
            visited[i] = 1;
            backtracking(k - dungeons[i][1], dungeons, visited, cnt + 1); // 소모 피로도
            visited[i] = 0;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    vector<int> visited(dungeons.size(), 0);
    backtracking(k, dungeons, visited, 0);
    return answer;
}

File: Programmers/Backtracking/피로도.py
answer = 0

def backtracking(dungeons, k, n, visited, cnt):
    global answer
    answer = max(answer, cnt)

    for i in range(n):
        if not visited[i] and k >= dungeons[i][0]:
            visited[i] = True
            backtracking(dungeons, k - dungeons[i][1], n, visited, cnt + 1)
            visited[i] = False


def solution(k, dungeons):
    global answer
    n = len(dungeons)
    visited = [False] * (n)
    backtracking(dungeons, k, n, visited, 0)
    return answer